### PR TITLE
Davins Gruppe (Sprint #1: Add Dropout Tutorial, Torch & Common Tests, and Flax Support for Ensemble)

### PR DESCRIPTION
**Pull request by D4V1ND**
## Issue

[#15](https://github.com/Cortys/SEP-Probly-WS2526/issues/15), [#14](https://github.com/Cortys/SEP-Probly-WS2526/issues/14), [#11](https://github.com/Cortys/SEP-Probly-WS2526/issues/11) and Tests for all Transformations
## Motivation and Context

- Added a Dropout tutorial notebook. 
- Added PyTorch tests for all transformations.
- Added common tests for all transformations. 
- Added Flax support for ensemble transformation.
- Added ensemble transformation tests for flax.

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?
- ran pytest locally

---

## Checklist

-   [x] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [x] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [ ] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: src/probly/transformation/ensemble/flax.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+from flax import nnx
+import jax
+import jax.numpy as jnp
+import numpy as np
+import warnings 
+warnings.filterwarnings("ignore")
+
+
+# In[141]:
+
```
Comment (Cortys) [inline src/probly/transformation/ensemble/flax.py]: 1. Please do not ignore warnings without a good reason.
2. Always provide a docstring at the top of each module.
3. Do not commit shebangs or `# In[xxx]` comments used during development.

```suggestion
"""TODO"""

from __future__ import annotations

from flax import nnx
import jax
import jax.numpy as jnp
import numpy as np
```

File: tests/probly/transformation/dropout/test_common.py [hunk 2]
```diff
@@ -20,3 +20,4 @@ def test_invalid_p_value(dummy_predictor: Predictor) -> None:
     p = 2
     with pytest.raises(ValueError, match=f"The probability p must be between 0 and 1, but got {p} instead."):
         dropout(dummy_predictor, p=p)
+
```
Comment (Cortys) [inline tests/probly/transformation/dropout/test_common.py]: Please do not add whitespace at the end of lines.
To prevent such issues in the future, remember to fix all Ruff warnings before creating a PR.

File: tests/probly/transformation/dropconnect/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+"""Tests for the dropconnect module"""
```
Comment (Cortys) [inline tests/probly/transformation/dropconnect/__init__.py]:
```suggestion
"""Tests for the dropconnect module."""
```

File: tests/probly/transformation/ensemble/test_Torch.py
Comment (Cortys) [inline tests/probly/transformation/ensemble/test_Torch.py]: Rename this file to `test_torch.py`.

File: tests/probly/transformation/bayesian/test_torch.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+from probly.transformation.bayesian import torch
+from probly.layers.torch import BayesLinear, BayesConv2d
+from tests.probly.torch_utils import count_layers
+from probly.transformation import bayesian
+
+from torch import nn
+
+
+class TestNetworkArchitectures:
```
Comment (Cortys) [inline tests/probly/transformation/bayesian/test_torch.py]: Overall this is a sound testing strategy!
To make the test suite more comprehensive, please also add tests that check whether the params (`use_base_weight`, `posterior_std`, `prior_mean` and `prior_std`) are correctly passed to the bayesian layers.

File: tests/probly/transformation/evidential/classification/test_torch.py
Comment (Cortys) [inline tests/probly/transformation/evidential/classification/test_torch.py]: Nice tests! For completeness, please also add a test for the `torch_conv_linear_model` fixture.

File: tests/probly/transformation/evidential/regression/test_torch.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+import pytest
+
+from probly.transformation.evidential.regression import evidential_regression
+from tests.probly.torch_utils import count_layers
+from probly.layers.torch import NormalInverseGammaLinear
+
+torch = pytest.importorskip("torch")
+
+from torch import nn  # noqa: E402
```
Comment (Cortys) [inline tests/probly/transformation/evidential/regression/test_torch.py]: Please make sure that all Ruff warnings are addressed.

```suggestion
"""TODO!"""

from __future__ import annotations

import pytest

from probly.layers.torch import NormalInverseGammaLinear
from probly.transformation.evidential.regression import evidential_regression
from tests.probly.torch_utils import count_layers

torch = pytest.importorskip("torch")

from torch import nn  # noqa: E402
```

File: tests/probly/transformation/dropconnect/test_torch.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        count_sequential_modified = count_layers(model, nn.Sequential)
+
+        assert model is not None
+        assert isinstance(model, type(torch_dropout_model))
+        assert count_drop_connect_modified == count_linear_original - 1
+        assert count_drop_connect_original == 0
+        assert count_linear_modified == 1
+        assert count_dropout_modified == count_dropout_original
+        assert count_sequential_original == count_sequential_modified
+
```
Comment (Cortys) [inline tests/probly/transformation/dropconnect/test_torch.py]: Please remove this test for now, as the intended semantics of the dropconnect transform in the presence of existing Dropout layers are not defined yet.

While this tests passes now, this might very well change once a clear strategy for this case has been decided upon.

File: src/probly/transformation/ensemble/flax.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    for attr_name in dir(model): 
+        attr = getattr(model, attr_name)
+        if isinstance(attr, nnx.Param): 
+            rng_key, subkey = jax.random.split(rng_key)
+            rngs = nnx.Rngs(params=subkey)
+            model.__dict__[attr_name] = nnx.Param(rngs.params.uniform(attr.value.shape))
+        elif isinstance(attr, nnx.Module): 
+            rng_key = reset_parameter(attr, rng_key)
+    return rng_key
```
[the remark this answers is not in the document]
Comment (D4V1ND) [reply src/probly/transformation/ensemble/flax.py]: hmm that means, the reset_parameter function must detect what kinds of layer each module is and applies the correct reset logic for that layer type?

as if I remove this function, than it's no longer an ensemble transformation no?
